resin/models/supervisor.py

```diff
@@ -0,0 +1,363 @@
+import os
+
+from ..base_request import BaseRequest
+from ..settings import Settings
+from .. import exceptions
+
+
+class Supervisor(object):
+    """
+    This class implements supervisor model for Resin Python SDK.
+
+    Attributes:
+        SUPERVISOR_API_VERSION (str): supervisor API version.
+        RESIN_SUPERVISOR_ADDRESS (str): supervisor endpoint address on device.
+        RESIN_SUPERVISOR_API_KEY (str): supervisor API key on device.
+        _on_device (bool): API endpoint flag.
+            If True then all commands will be sent to the API on device.
+            If False then all command will be sent to the Resin API proxy endpoint (api.resin.io/supervisor/<url>).
+            If RESIN_SUPERVISOR_ADDRESS and RESIN_SUPERVISOR_API_KEY are available, _on_device will be set to True by default. Otherwise, it's False.
+
+    """
+
+    SUPERVISOR_API_VERSION = 'v1'
+    RESIN_SUPERVISOR_ADDRESS = os.environ.get('RESIN_SUPERVISOR_ADDRESS')
+    RESIN_SUPERVISOR_API_KEY = os.environ.get('RESIN_SUPERVISOR_API_KEY')
+
+    # _on_device = True, if RESIN_SUPERVISOR_ADDRESS and RESIN_SUPERVISOR_API_KEY env vars are avaiable => can communicate with both supervisor API endpoints on device or Resin API endpoints.
+    # _on_device = False, if RESIN_SUPERVISOR_ADDRESS and RESIN_SUPERVISOR_API_KEY env vars are not avaiable => can only communicate with Resin API endpoints.
+    if RESIN_SUPERVISOR_ADDRESS is None or RESIN_SUPERVISOR_API_KEY is None:
+        _on_device = False
+    else:
+        _on_device = True
+
+    def __init__(self):
+        self.base_request = BaseRequest()
+        self.settings = Settings()
+
+    def _check_args(self, device_id, app_id):
+        if device_id is None:
+            exceptions.MissingOption('device_id')
+
+        if app_id is None:
+            exceptions.MissingOption('app_id')
+
+    def force_api_endpoint(self, endpoint):
+        """
+        Force all API commands to a specific endpoint.
+
+        Args:
+            endpoint (bool): True if selecting the API on device. False if selecting the Resin API proxy endpoint.
+
+        Raises:
+            InvalidOption: if endpoint value is not bool.
+
+        """
+
+        try:
+            self._on_device = bool(endpoint)
+        except ValueError:
+            exceptions.InvalidOption('endpoint')
+
+    def ping(self, device_id=None, app_id=None):
+        """
+        Check that the supervisor is alive and well.
+        No need to set device_id and app_id if command is sent to the API on device.
+
+        Args:
+            device_id (Optional[str]): device id.
+            app_id (Optional[str]): application id.
+
+        Returns:
+            str: `OK` signals that the supervisor is alive and well.
+
+        Raises:
+            InvalidOption: if the endpoint is Resin API proxy endpoint and device_id or app_id is not set.
+
+        """
+
+        if not self._on_device:
```

Comment:
You are repeating the "if self._on_device" check every time, this could be abstracted with a function like:
```python
def do_request(self, endpoint, data, device_id=None, app_id=None, method='POST'):
	if not self._on_device:
		self._check_args(device_id, app_id)
		data = {
			'deviceId': device_id,
			'appId': app_id,
			'method': method,
			'data': data
		}
		return self.base_request.request(
			'supervisor/' + endpoint,
			'POST',
			data=data,
			endpoint=self.settings.get('api_endpoint'),
			login=True
		)
	else:
		return self.base_request.request(
			endpoint,
			method,
			endpoint=self.RESIN_SUPERVISOR_ADDRESS,
			api_key=self.RESIN_SUPERVISOR_API_KEY
		)
```

This would remove around 20 lines of duplicated code on every API-calling method. I will approve if you're convinced this doesn't make it more elegant, but I strongly suggest DRYing it up.